test/op_shape_test.cpp

```diff
@@ -2303,6 +2303,65 @@ TEST_CASE(slice_shape)
     expect_shape(migraphx::shape{migraphx::shape::int32_type, {2, 2, 1}, {6, 3, 1}},
                  migraphx::make_op("slice", {{"axes", {2}}, {"starts", {2}}, {"ends", {10}}}),
                  input);
+    expect_shape(migraphx::shape{migraphx::shape::int32_type, {2, 2, 1}, {6, 3, 1}},
+                 migraphx::make_op("slice", {{"axes", {2}}, {"starts", {-1}}, {"ends", {10}}}),
+                 input);
+}
+
+TEST_CASE(slice_dyn_shape0)
+{
+    migraphx::shape input{migraphx::shape::int32_type, {{2, 3, 0}, {7, 8, 0}, {2, 3, 0}}};
+
+    // Slice axis 1 to size 4-1=3
+    expect_shape(migraphx::shape{migraphx::shape::int32_type, {{2, 3, 0}, {3, 3, 0}, {2, 3, 0}}},
+                 migraphx::make_op("slice", {{"axes", {1}}, {"starts", {1}}, {"ends", {4}}}),
+                 input);
+}
+
+TEST_CASE(slice_dyn_shape1)
+{
+    migraphx::shape input{migraphx::shape::int32_type, {{2, 3, 0}, {7, 8, 0}, {2, 3, 0}}};
+    // Slice axis 1 to size 4-1=3, negative index
+    expect_shape(migraphx::shape{migraphx::shape::int32_type, {{2, 3, 0}, {3, 3, 0}, {2, 3, 0}}},
+                 migraphx::make_op("slice", {{"axes", {1}}, {"starts", {1}}, {"ends", {-4}}}),
+                 input);
```

Comment:
Doesn't look right. -4 index for a 7 length vector is index 3. So `start = 1` and `end = 3`. Since the end, is exclusive the slice  should have length 2.

Reply:
You are correct.  This test is giving the wrong result.  The underlying cause is that negative values in the attribute `ends` aren't being passed correctly.  This debug output shows that when values of {-4, -3} were set, the array received by the operator was [4, 0]:  (ignore other incorrect values)

`bpickrel@home-tower:~/AMDMIGraphX/build$ bin/test_op_shape_test slice_dyn_shape1
[   RUN    ] slice_dyn_shape1
ends:  4 0 
ends:  4 0 
FAILED: Incorrect shape for slice[axes={1, 2},starts={1, 1},ends={-4, -3}]: dynamic, int32_type, {[2, 3, 0], [2, 3, 0], [2, 3, 0]} != dynamic, int32_type, {[2, 3, 0], [3, 3, 0], [2, 3, 0]}
    dynamic, int32_type, {[2, 3, 0], [7, 8, 0], [2, 3, 0]}
[ COMPLETE ] slice_dyn_shape1`

Reply:
Learned that negative inputs are being intercepted by the normalize_attributes op.  We'd like the allowed range to be [-L, L] where L is the size of the N'th dimension.  Unfortunately L needs two different values for max and min lengths, and normalize_attributes doesn't (yet) handle dynamic dimensions.  I think it won't handle L having different values for different axis numbers either.

Options for dealing with this are:

- cut "starts" and "ends" out of the attribute() method code and handle range checking explicitly in normalize_compute_shape().  This makes sense if the requirements are specific to this op. only and generic normalize_axes handling shouldn't be expected to handle them (is this true?)
- add dynamic shape support to normalize_axes .
- don't support negative length inputs in the Slice operator for now; wait until normalize_axes is updated as a separate task, and remove test case with negative input for now.

Reply:
Another issue that came up during debugging:  our notation can't handle slices that include the end of a non-fixed dynamic dimension array.  Example:  suppose array A has size 5 (indexes from 0 to 4).  In Numpy, A[1:] is a slice from the 1th to last element, slicing off the 0'th element.  We don't support leaving a value blank in "ends", so the only way to make this slice is with starts={1}, ends={5}.  But if we're slicing an array with dynamic dimension with a min and max length, no constant "ends" value will fit both.  Note that negative indexing doesn't solve this, because ends={-1} refers to A[4] and the notation requires that the index be one greater than the last element you want--in other words, A[1 : -1] gives a slice with elements A(1), A(2), and A(3) only.

Reply:
`start`, `end` and `axes` are fixed literal  values. `normalize_attributes` is only called on those three attributes. 
Therefore, it doesn't need to handle dynamic shapes.

Reply:
> But if we're slicing an array with dynamic dimension with a min and max length, no constant "ends" value will fit both
Wouldn't `end[i]` value equal to `max_lens[i]` where `i` is axis, work ?